src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/client/VisitDetailsClient.kt

```diff
@@ -0,0 +1,53 @@
+package uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.client
+
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.hmpps.auth.UserDetails
+import uk.gov.justice.digital.hmpps.hmppsmanageprisonvisitsorchestration.dto.visit.scheduler.VisitDto
+import java.time.Duration
+import java.util.Optional
+
+@Component
+class VisitDetailsClient(
+  private val visitSchedulerClient: VisitSchedulerClient,
+  private val hmppsAuthClient: HmppsAuthClient,
+  @Value("\${hmpps.auth.timeout:10s}") private val apiTimeout: Duration,
+) {
+  companion object {
+    const val NOT_KNOWN = "NOT_KNOWN"
+  }
+  fun getFullVisitDetailsByReference(
+    reference: String,
+  ): VisitDto? {
+    val visitDto = visitSchedulerClient.getVisitByReference(reference)
+    visitDto?.also {
+      val createdByMono = getUserDetails(visitDto.createdBy)
+      val updatedByMono = getUserDetails(visitDto.updatedBy)
+      val cancelledByMono = getUserDetails(visitDto.cancelledBy)
+
```

Comment:
wonder if you could do something clever here and if the visitDto.createdBy is the same as visitDto.updatedBy or the visitDto.cancelledBy only call it once?

```
val createdByMono = getUserDetails(visitDto.createdBy)
val updatedByMono = if (visitDto.createdBy==visitDto.updatedBy) createdByMono else getUserDetails(visitDto.updatedBy)
```
not sure how that would work with the Mono zip though!
Just i'm concerned about Auth and it's performance.

Reply:
I will explore moving usernames to a Set and making fewer calls.

Reply:
raised a story to pick these up later - https://dsdmoj.atlassian.net/browse/VB-2045